uvicorn/server.py

```diff
@@ -277,6 +279,28 @@ async def shutdown(self, sockets: Optional[List[socket.socket]] = None) -> None:
             connection.shutdown()
         await asyncio.sleep(0.1)
 
+        # When 3.10 is not supported anymore, use `async with asyncio.timeout(...):`.
+        try:
+            await asyncio.wait_for(
+                self._wait_tasks_to_complete(),
+                timeout=self.config.timeout_graceful_shutdown,
+            )
+        except TimeoutError:
+            logger.error(
+                "Cancel %s running task(s), timeout_graceful_shutdown exceeded",
+                len(self.server_state.tasks),
+            )
+            for t in self.server_state.tasks:
+                if sys.version_info < (3, 9):
+                    t.cancel()
+                else:
+                    t.cancel(msg="Task cancelled, timeout_graceful_shutdown exceeded")
```

Comment:
I discussed handling `self.server_state.connections`, but this has already been before the timeout through
```python
for connection in list(self.server_state.connections):
    connection.shutdown()
```

Handling tasks should be sufficient, as proven in my test.